azurelinuxagent/common/cgroups.py

```diff
@@ -0,0 +1,689 @@
+                else:
+                    logger.info("Daemon process pid {0} cgroup managed by systemd".format(pid))
+            cgroups_enabled = True
+            status = "OK"
+        except CGroupsException as cge:
+            status = cge.msg
+
+        from azurelinuxagent.common.event import add_event, WALAEventOperation
+        add_event(
+            AGENT_NAME,
+            version=CURRENT_VERSION,
+            op=WALAEventOperation.InitializeCGroups,
+            is_success=cgroups_enabled,
+            message=status,
+            log_event=False)
+
+    @staticmethod
+    def add_to_extension_cgroup(name, pid=int(os.getpid())):
+        """
+        Create cgroup directories for this extension in each of the hierarchies and add this process to the new cgroup.
+        Should only be called when creating sub-processes and invoked inside the fork/exec window. As a result,
+        there's no point in returning the CGroups object itself; the goal is to move the child process into the
+        cgroup before the new code even starts running.
+
+        :param str name: Short name of extension, suitable for naming directories in the filesystem
+        :param int pid: Process id of extension to be added to the cgroup
+        """
+        if not CGroups.enabled():
+            return
+        if name == AGENT_NAME:
+            logger.warn('Extension cgroup name cannot match agent cgroup name ({0})'.format(AGENT_NAME))
+            return
+
+        try:
+            logger.info("Move process {0} into cgroups for extension {1}".format(pid, name))
+            CGroups.for_extension(name).add(pid)
+        except Exception as ex:
+            logger.warn("Unable to move process {0} into cgroups for extension {1}: {2}".format(pid, name, ex))
+
+    @staticmethod
+    def get_my_cgroup_path(hierarchy_id):
+        """
+        Get the cgroup path "suffix" for this process for the given hierarchy ID. The leading "/" is always stripped,
+        so the suffix is suitable for passing to os.path.join(). (If the process is in the root cgroup, an empty
+        string is returned, and os.path.join() will still do the right thing.)
+
+        :param hierarchy_id: str
+        :return: str
+        """
+        cgroup_paths = fileutil.read_file("/proc/self/cgroup")
+        for entry in cgroup_paths.splitlines():
+            fields = entry.split(':')
+            if fields[0] == hierarchy_id:
+                return fields[2].lstrip(os.path.sep)
+        raise CGroupsException("This process belongs to no cgroup for hierarchy ID {0}".format(hierarchy_id))
+
+    @staticmethod
+    def get_hierarchy_id(hierarchy):
+        """
+        Get the cgroups hierarchy ID for a given hierarchy name
+
+        :param hierarchy:
+        :return: str
+        """
+        cgroup_states = fileutil.read_file("/proc/cgroups")
+        for entry in cgroup_states.splitlines():
+            fields = entry.split('\t')
```

Comment:
Thank you CGroups for using '\t' and ':' for separation.

Reply:
Please consider adding a comment on what you're parsing.

```text
#subsys_name    hierarchy       num_cgroups     enabled
cpuset  9       1       1
cpu     3       74      1
cpuacct 3       74      1
blkio   10      74      1
memory  12      111     1
devices 6       74      1
freezer 11      1       1
net_cls 2       1       1
perf_event      8       1       1
net_prio        2       1       1
hugetlb 5       1       1
pids    4       75      1
rdma    7       1       1
```

Please consider returning an object with each field called out to make it easier for readers in a future PR.